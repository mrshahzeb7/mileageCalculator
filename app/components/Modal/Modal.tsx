import React, {
  forwardRef,
  Ref,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import {
  ModalProps,
  Modal as RNModal,
  StyleProp,
  StyleSheet,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';
import {isIos, responsiveHeight, responsiveWidth} from '../../utils';
import {colors} from '../../theme';

interface Props extends ModalProps {
  contentStyle?: StyleProp<ViewStyle>;
  closeModal?(): any;
  statusBarTranslucent?: boolean;
  style?: StyleProp<ViewStyle>;
  onRequestClose?(): void;
}

export interface ModalRef {
  toggleModal?(): void;
  openModal?(): void;
  close?(): void;
  setVisibility?(value: boolean): void;
}

export const Modal = forwardRef(function Modal(
  props: Props,
  ref: Ref<ModalRef>,
) {
  const {
    children,
    style,
    contentStyle,
    closeModal,
    onRequestClose,
    ...restProps
  } = props;

  const [modalVisible, setModalVisible]: any = useState(false);

  const toggleModal = () => {
    setModalVisible((p: boolean) => !p);
  };
  const setVisibility = (value: boolean) => {
    setModalVisible(value);
  };

  const close = () => {
    setModalVisible(false);
  };

  useImperativeHandle(ref, () => ({
    close,
    setVisibility,
  }));

  const onCloseModal = () => {
    onRequestClose?.();
    closeModal?.();
    close();
  };

  return (
    <RNModal
      visible={modalVisible}
      onRequestClose={onCloseModal}
      transparent
      statusBarTranslucent={true}
      {...restProps}>
      <View style={[styles.container, style]}>
        <TouchableOpacity
          style={styles.touchableArea}
          onPress={onRequestClose || toggleModal}
        />
        <View style={[styles.modalContentContainer, contentStyle]}>
          {children}
        </View>
      </View>
    </RNModal>
  );
});

export function useModalRef() {
  const modalRef = useRef<ModalRef>(null);
  return modalRef;
}

Modal.defaultProps = {
  animationType: 'fade',
  statusBarTranslucent: true,
};

const styles = StyleSheet.create({
  container: {
    width: responsiveWidth(100),
    ...(isIos
      ? {flex: 1}
      : {height: responsiveHeight(100) + responsiveHeight(5)}),
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.modalOverlay,
  },
  touchableArea: {
    position: 'absolute',
    flex: 1,
    height: responsiveHeight(110),
    width: responsiveWidth(100),
  },
  modalContentContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
});
