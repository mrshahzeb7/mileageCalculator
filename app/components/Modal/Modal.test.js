import 'react-native';
import React from 'react';
import {Modal} from './Modal';
import renderer from 'react-test-renderer';

describe('Modal', () => {
  it(' Modal renders correctly', () => {
    const tree = renderer.create(<Modal />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
