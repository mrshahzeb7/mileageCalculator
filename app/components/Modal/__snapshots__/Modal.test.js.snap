// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal  Modal renders correctly 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  statusBarTranslucent={true}
  transparent={true}
  visible={false}
/>
`;
